covid_cases/tasks.py

```diff
@@ -0,0 +1,145 @@
+import re
+from datetime import date
+from functools import lru_cache
+
+import requests
+from celery.exceptions import SoftTimeLimitExceeded
+from django.conf import settings
+from django.db import transaction
+from django.db.models import Max, Sum
+from requests.exceptions import RequestException
+
+from covid_cases.models import District, Province, SubDistrict, Ward, WardCase
+from healthcheck.celery import app
+
+
+def normalise_text(text: str) -> str:
+    return re.sub(r"\s+", " ", text).strip().title()
+
+
+def get_api_total_cases(api_data: dict) -> int:
+    total_cases = 0
+    for record in api_data["features"]:
+        total_cases += record["attributes"]["Tot_No_of_Cases"]
+    return total_cases
+
+
+def get_database_total_cases() -> int:
+    latest_date = WardCase.objects.aggregate(latest_date=Max("date"))["latest_date"]
+    return (
+        WardCase.objects.filter(date=latest_date).aggregate(
+            total=Sum("total_number_of_cases")
+        )["total"]
+        or 0
+    )
+
+
+@lru_cache(maxsize=None)
+def get_province(province: str):
+    province, _ = Province.objects.get_or_create(name=province)
+    return province
+
+
+@lru_cache(maxsize=None)
+def get_district(province: str, district: str):
+    province = get_province(province)
+    district, _ = District.objects.get_or_create(province=province, name=district)
+    return district
+
+
+@lru_cache(maxsize=None)
+def get_sub_district(
+    province: str, district: str, sub_district: str, sub_district_id: int
+):
+    district = get_district(province, district)
+    sub_district, _ = SubDistrict.objects.get_or_create(
+        district=district, name=sub_district, subdistrict_id=sub_district_id
+    )
+    return sub_district
+
+
+def get_ward(
+    province: str,
+    district: str,
+    sub_district: str,
+    sub_district_id: int,
+    ward_id: str,
+    ward_number: str,
+) -> Ward:
+    sub_district = get_sub_district(province, district, sub_district, sub_district_id)
+    ward, _ = Ward.objects.get_or_create(
+        sub_district=sub_district, ward_id=ward_id, ward_number=ward_number
+    )
+    return ward
```

Comment:
These might make more sense in a utils.py file

Reply:
Maybe on the models in models.py instead? Since it's closely tied to the models

Reply:
Agreed!